### PR TITLE
Autotuner for int mm Triton kernels

### PR DESCRIPTION
**Pull request by cpuhrsch**
```
[ 8:58PM (nightly20240311py310) /scratch/cpuhrsch/dev/ao/benchmarks - git:intmmbenchmarks1]$ TORCHAO_AUTOTUNER_ENABLE=0 python intmm.py sam_shapes.csv
fn,m,k,n,fp_time,int_mm_time,ratio
<function run_int_mm_benchmark at 0x7f42252f29e0>,32768,3072,768,0.7926271820068359,3.6111566162109376,0.219493992159917
<function run_int_mm_benchmark at 0x7f42252f29e0>,32768,768,2304,0.4677939224243164,2.7787774658203124,0.16834522669710103
<function run_int_mm_benchmark at 0x7f42252f29e0>,32768,768,3072,0.6118912124633789,3.285841979980469,0.1862205231387956
<function run_int_mm_benchmark at 0x7f42252f29e0>,32768,768,768,0.17190912246704101,1.7368576049804687,0.09897709632274326
<function run_int_mm_benchmark at 0x7f42252f29e0>,39200,768,2304,0.579788818359375,3.1330712890625,0.1850544609002062
<function run_int_mm_benchmark at 0x7f42252f29e0>,39200,768,768,0.20057088851928712,1.8391961669921875,0.10905355944020902
<function run_int_scaled_mm_benchmark at 0x7f430a506b00>,32768,3072,768,0.7390310668945312,3.5631512451171874,0.20740940141322164
<function run_int_scaled_mm_benchmark at 0x7f430a506b00>,32768,768,2304,0.8094003295898438,3.315210266113281,0.24414750939426128
<function run_int_scaled_mm_benchmark at 0x7f430a506b00>,32768,768,3072,1.0603622436523437,4.030166931152344,0.2631062836271039
<function run_int_scaled_mm_benchmark at 0x7f430a506b00>,32768,768,768,0.28380159378051756,1.9123712158203126,0.14840298339189378
<function run_int_scaled_mm_benchmark at 0x7f430a506b00>,39200,768,2304,0.9600614166259765,3.763394470214844,0.2551051781109645
<function run_int_scaled_mm_benchmark at 0x7f430a506b00>,39200,768,768,0.3342233657836914,2.057533416748047,0.16243885181312626
[ 8:58PM (nightly20240311py310) /scratch/cpuhrsch/dev/ao/benchmarks - git:intmmbenchmarks1]$ TORCHAO_AUTOTUNER_ENABLE=1 python intmm.py sam_shapes.csv
fn,m,k,n,fp_time,int_mm_time,ratio
INFO:root:Trying to load configs for NVIDIA A100-SXM4-40GB from /scratch/cpuhrsch/dev/ao/torchao/kernel/configs/data_a100.pkl
INFO:root:Loading best configs from file /scratch/cpuhrsch/dev/ao/torchao/kernel/configs/data_a100.pkl
<function run_int_mm_benchmark at 0x7fbfe839a9e0>,32768,3072,768,0.7970508575439453,0.39620609283447267,2.0117077247394524
<function run_int_mm_benchmark at 0x7fbfe839a9e0>,32768,768,2304,0.5419622421264648,0.41020416259765624,1.32120122500571
<function run_int_mm_benchmark at 0x7fbfe839a9e0>,32768,768,3072,0.7071027374267578,0.5336064147949219,1.3251391246833395
<function run_int_mm_benchmark at 0x7fbfe839a9e0>,32768,768,768,0.18099199295043944,0.13851648330688476,1.3066458852369918
<function run_int_mm_benchmark at 0x7fbfe839a9e0>,39200,768,2304,0.5925785446166992,0.45351936340332033,1.306622368160523
<function run_int_mm_benchmark at 0x7fbfe839a9e0>,39200,768,768,0.2048409652709961,0.1618739128112793,1.26543531143162
<function run_int_scaled_mm_benchmark at 0x7fc0cd6bab00>,32768,3072,768,0.7551487731933594,0.36495361328125,2.069163712078135
<function run_int_scaled_mm_benchmark at 0x7fc0cd6bab00>,32768,768,2304,0.8089702606201172,0.4003123092651367,2.0208478278001594
<function run_int_scaled_mm_benchmark at 0x7fc0cd6bab00>,32768,768,3072,1.0633113861083985,0.530063362121582,2.006008077699405
<function run_int_scaled_mm_benchmark at 0x7fc0cd6bab00>,32768,768,768,0.28652544021606446,0.17480703353881835,1.6390956039674311
<function run_int_scaled_mm_benchmark at 0x7fc0cd6bab00>,39200,768,2304,0.9518489837646484,0.4832358551025391,1.9697399804132356
<function run_int_scaled_mm_benchmark at 0x7fc0cd6bab00>,39200,768,768,0.33665023803710936,0.17441791534423828,1.9301356593597783
```

### CODE REVIEW

**Review by drisspg — APPROVED** [review]

File: torchao/kernel/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+### How to contribute
+
+### Environment variables
+
+`TORCHAO_AUTOTUNER_ENABLE=0`
+
+Set this to a nonzero value to enable the kernels generated by the autotuner. This is turned off by default, because it is still an experimental feature and also can take a long time to run.
+
+Searching a new config can take a long time and we'll save the updated data in `data.pkl`. If you'd like to contributed updated configs for your hardware or shapes, please open a pull request.
```
Comment (msaroufim) [inline torchao/kernel/README.md:13]: presumably people won't contribute the pickle file since that's not human readable? Also kind of a security issues for us to host pickle files

File: benchmarks/intmm.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+torch._dynamo.config.cache_size_limit = 128
+torch._dynamo.config.accumulated_cache_size_limit = 128
+
+dtype = torch.float16
+device = "cuda"
+
+
+
+def benchmark_in_ms(warmup, iters, f, *args, **kwargs):
```
Comment (msaroufim) [inline benchmarks/intmm.py:19]: put this in benchmark util instead?

File: benchmarks/intmm_shapes.csv [hunk 3]
```diff
@@ -0,0 +1,127 @@
+m,k,n
+1024,1024,2304
```
Comment (msaroufim) [inline benchmarks/intmm_shapes.csv:2]: where are these shapes coming from btw?

File: benchmarks/sam_shapes.csv [hunk 4]
```diff
@@ -0,0 +1,7 @@
+m,k,n
```
Comment (msaroufim) [inline benchmarks/sam_shapes.csv:1]: presumably you mean shapes of matmuls in sam?

File: test/kernel/test_autotuner.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        os.environ["TORCHAO_AUTOTUNER_ENABLE"] = "1"
+
+    def tearDown(self):
+        del os.environ["TORCHAO_AUTOTUNER_ENABLE"]
+
+    @parameterized.expand(
+            [
+                ("cuda", torch.bfloat16),
+                ("cuda", torch.bfloat16),
+                # ("cpu", torch.bfloat16),
```
Comment (msaroufim) [inline test/kernel/test_autotuner.py:30]: nit: remove comments

File: test/kernel/test_autotuner.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                ("cuda", torch.float16),
+                ("cuda", torch.float16),
+                 #("cpu", torch.float16),
+            ])
+    def test_int_mm(self, device, dtype):
+        from torchao.kernel import intmm_triton
+        logging.basicConfig(level=logging.INFO)
```
Comment (msaroufim) [inline test/kernel/test_autotuner.py:37]: set logging at top of file

File: torchao/kernel/README.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
+## Autotuner and custom Triton kernels
+
+### Use case
```
Comment (msaroufim) [inline torchao/kernel/README.md:3]: Is intent to fill this out later? Might be better to open an issue if you'd like to do this later

File: torchao/kernel/autotuner.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+def do_bench_triton(fn, warmup=25, rep=100, grad_to_none=None, quantiles=None, fast_flush=True, return_mode="mean"):
+    assert return_mode in ["min", "max", "mean", "median"]
+    import torch
+    """
+    Benchmark the runtime of the provided function. By default, return the median runtime of :code:`fn` along with
+    the 20-th and 80-th performance percentile.
+
+    :param fn: Function to benchmark
+    :type fn: Callable
+    :param warmup: Warmup time (in ms)
```
Comment (msaroufim) [inline torchao/kernel/autotuner.py:18]: warmup is not a time it seems to be an int so not in ms

File: torchao/kernel/autotuner.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+    if quantiles is not None:
+        ret = torch.quantile(times, torch.tensor(quantiles, dtype=torch.float)).tolist()
+        if len(ret) == 1:
+            ret = ret[0]
+        return ret
+    return getattr(torch, return_mode)(times).item()
+
+BEST_CONFIGS = None
+
+AUTOTUNER_DATA_PATH = os.getenv('TORCHAO_AUTOTUNER_DATA_PATH', None)
```
Comment (msaroufim) [inline torchao/kernel/autotuner.py:86]: put all global variables at the top of the file so they're easier to find

File: torchao/kernel/intmm_triton.py [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+            (i, j, k, 6, 8),
+            (i, j, k, 7, 8),
+            (i, j, k, 8, 8),
+        ] for 
+        (i, j, k) in 
+        itertools.product(int8_powers_of_two,
+                          int8_powers_of_two,
+                          int8_powers_of_two)], [])
+
+# int8_mm_kernel_configs = [
```
Comment (msaroufim) [inline torchao/kernel/intmm_triton.py:38]: delete?

File: torchao/kernel/intmm_triton.py [hunk 11]
```diff
@@ -0,0 +1,349 @@
+import torch
+from torchao.kernel.autotuner import get_best_config_fn
+from torch._higher_order_ops.out_dtype import out_dtype
+from torch._dynamo import is_compiling as dynamo_is_compiling
+
+import triton
+import triton.language as tl
+import itertools
+import os
+int8_powers_of_two = [32, 64, 128, 256]
```
Comment (msaroufim) [inline torchao/kernel/intmm_triton.py:10]: do you envision people wanting to add more options here and for int8 kernel configs?

**Review by msaroufim — CHANGES_REQUESTED** [review]
some nits

File: test/kernel/test_autotuner.py (hunk 5 above)
Comment (cpuhrsch) [reply test/kernel/test_autotuner.py:30]: I'll turn those into TODOs. It should also work on CPU.

File: benchmarks/sam_shapes.csv (hunk 4 above)
Comment (cpuhrsch) [reply benchmarks/sam_shapes.csv:1]: Yes, SAM vit_b batch size 16 to be precise

File: torchao/kernel/README.md (hunk 1 above)
Comment (cpuhrsch) [reply torchao/kernel/README.md:13]: I added https://github.com/pytorch-labs/ao/pull/41/files#diff-4986e1d3257adc0a73b17fd6f21ef9d3b2c0eaec9027381e6f2de89e5be0e6b5 to make it easier to inspect. It stores the triton Configs, so it's a bit more difficult to make them human readable by default.

**Review by msaroufim — COMMENTED** [review]
Unblocking for now

File: benchmarks/intmm.py (hunk 2 above)
Comment (cpuhrsch) [reply benchmarks/intmm.py:19]: Once I add the next benchmark for weight only

File: torchao/kernel/autotuner.py (hunk 8 above)
Comment (cpuhrsch) [reply torchao/kernel/autotuner.py:18]: Why can't milliseconds be given in int?

Comment (cpuhrsch) [reply torchao/kernel/autotuner.py:18]: I think this is just saying "run warmup until at least 25ms were spent"

File: torchao/kernel/autotuner.py (hunk 9 above)
Comment (cpuhrsch) [reply torchao/kernel/autotuner.py:86]: Good point

File: torchao/kernel/intmm_triton.py (hunk 10 above)
Comment (cpuhrsch) [reply torchao/kernel/intmm_triton.py:38]: I wanted to leave these as reference from core. I'll add a comment.

File: torchao/kernel/intmm_triton.py (hunk 11 above)
Comment (cpuhrsch) [reply torchao/kernel/intmm_triton.py:10]: Eventually, yes. Follow up here includes making it more extensible for other kernels. Adding support for mixed precision should help that.
